pkg/messaging/mqtt/subscriber.go

```diff
@@ -18,55 +18,99 @@ import (
 var (
 	errSubscribeTimeout   = errors.New("failed to subscribe due to timeout reached")
 	errUnsubscribeTimeout = errors.New("failed to unsubscribe due to timeout reached")
+	errAlreadySubscribed  = errors.New("already subscribed to topic")
+	errNotSubscribed      = errors.New("not subscribed")
+	errEmptyTopic         = errors.New("empty topic")
+	errEmptyID            = errors.New("empty ID")
 )
 
 var _ messaging.Subscriber = (*subscriber)(nil)
 
 type subscriber struct {
-	client  mqtt.Client
-	timeout time.Duration
-	logger  log.Logger
+	address       string
+	timeout       time.Duration
+	logger        log.Logger
+	subscriptions map[string]map[string]mqtt.Client
 }
 
 // NewSubscriber returns a new MQTT message subscriber.
 func NewSubscriber(address string, timeout time.Duration, logger log.Logger) (messaging.Subscriber, error) {
-	client, err := newClient(address, timeout)
-	if err != nil {
-		return nil, err
-	}
-
 	ret := subscriber{
-		client:  client,
-		timeout: timeout,
-		logger:  logger,
+		address:       address,
+		timeout:       timeout,
+		logger:        logger,
+		subscriptions: make(map[string]map[string]mqtt.Client),
 	}
 	return ret, nil
 }
 
-func (sub subscriber) Subscribe(topic string, handler messaging.MessageHandler) error {
-	token := sub.client.Subscribe(topic, qos, sub.mqttHandler(handler))
+func (sub subscriber) Subscribe(id, topic string, handler messaging.MessageHandler) error {
+	if id == "" {
+		return errEmptyID
+	}
+	if topic == "" {
+		return errEmptyTopic
+	}
+	// Check topic
+	s, ok := sub.subscriptions[topic]
+	switch ok {
+	case true:
+		// Check client ID
+		if _, ok := s[id]; ok {
+			return errAlreadySubscribed
+		}
+	default:
```

Comment:
There is no need to create a new client for each topic, but only for each client ID. We have to store topics for each client anyway (because the library does not provide us to access to all the subscribed topics for a single client) so that we know when to close the client (when we empty all the topics the client is subscribed to), we can do something like this:
```go

type subscription struct {
	client mqtt.Client
	topics []string
}

// NewSubscriber returns a new MQTT message subscriber.
func NewSubscriber(address string, timeout time.Duration, logger log.Logger) (messaging.Subscriber, error) {
	ret := subscriber{
		address:       address,
		timeout:       timeout,
		logger:        logger,
		subscriptions: make(map[string]subscription),
	}
	return ret, nil
}
```

Reply:
Have a look at the implementation. With topics being a map, we would have a simpler implementation and we wouldn't need to write the `contains` and `delete` functions